### DISCUSSION TITLE
Wrapping packets in custom header

### DISCUSSION THREAD

**Discussion by ahura-proxima** [nasa/fprime#2175] [Q&A] [answered]
Is there a built-in way to wrap packets from Packets.xml in your own custom header before sending to gds? Also are there any docs where I can get more clarity on what the packets from the fprime deployment look like being sent and what's happening on the gds side to parse them and break up the channels for display?

**Comment by Joshua-Anderson** [comment]
@ahura-proxima No documentation yet, but the code doing packet processing is [here](https://github.com/fprime-community/fprime-gds/blob/v3.3.0/src/fprime_gds/common/decoders/pkt_decoder.py)

Custom packet headers are possible, but you'll need to customize the GDS to parse them.

Long term roadmap includes adding support for the CCSDS Space Packet Protocol, which will likely require updating the GDS packet processing to be more easily customizable in the future.

I think the fprime packet format is totally undocumented at the moment, so if you think better documentation there would aid your project you should open a issue requesting adding some.

**Reply by ahura-proxima** [reply]
Thank you! Got it, is there a recommended form of going about implementing custom headers? For example, something like the message pre-hooks for the comm send buffer where I could wrap packets on the way out?

**Reply by Joshua-Anderson** [reply, marked as answer]
You can implement your own [framing protocol](https://github.com/nasa/fprime/tree/v3.3.2/Svc/FramingProtocol), which is then used by the F' packet framer and deframer components.
